### DISCUSSION TITLE
TVB GSOC 2024

### DISCUSSION THREAD

**Discussion by liadomide** [the-virtual-brain/tvb-root#699] [Ideas]
## The Virtual Brain: An open-source simulator for whole brain network modeling.

Simulating the human brain is the holy grail of neuroscience – offering a pioneering tool for understanding how our brain works and how to deal with its disorders like stroke, epilepsy or neurodegenerative diseases like Alzheimer’s or Parkinson’s.

While large-scale research initiatives simulate neurons and small brain regions at the cellular level on massively parallel hardware, they are still years away from clinical applications.

The Virtual Brain ([TVB](https://www.thevirtualbrain.org/)) takes a different approach and reduces complexity on the micro level to attain the macro organization: a TVB model of a patient's brain generates sufficiently accurate EEG, MEG, BOLD and SEEG signals by reducing the complexity millionfold through methods from statistical physics. The key is TVB’s hybrid approach of merging individual anatomy from brain imaging data with state-of-the-art mathematical modeling.

### [1] 3D Head widget(s)

**Description:** 
In the TVB (https://www.thevirtualbrain.org/) ecosystem there is a new code repository called [tvb-widgets](https://github.com/the-virtual-brain/tvb-widgets) offering neat UI widgets for Jupyterlab environments. These widgets are compatible with TVB data formats and able to display these data in different forms: 2D, 3D. 
The purpose of this project is to implement a new set of widgets, which would allow users to work  in Jupyter Lab in 3D, with the connectivity matrices (nodes & edges), surfaces (cortical, subcortical, face), sensors involved in a TVB simulation. 
Necessary features for this widget: select from a drive the list of available datesets, load/unload a dataset, display connectivity matrix, normalize matrix, resect connections, resect nodes, change connection weights, save resulted data, change colors of the display, move in 3D, zoom, etc. Of course, these new widgets have to run in a Jupyterlab notebook as well.
Finally, it would be great to have all the widgets linked into the [tvb-ext-xircuits](https://github.com/the-virtual-brain/tvb-ext-xircuits) repository which is a Jupyterlab extension based on React JS. At the moment, only the [PhasePlaneWidget](https://drive.google.com/drive/folders/1DTMBSxbIcNdBBkSGCIEoyGqg3EIT4XDz) is linked there, but the rest could be added in a similar manner.

Examples of TVB data formats can be found on [Zenodo](https://zenodo.org/records/10128131). Connectivity matrices are available there as well. 
Check out our Jupyter [notebooks](https://github.com/the-virtual-brain/tvb-widgets/tree/main/notebooks) to play with the widgets we have available so far.

**Expected results:** A set of classes , with at least one demo Jupyter notebook, and unit tests.

**Preferred Tech keywords:** Python, IPywidgets, React JS, Jupyterlab, Jupyterlab extensions

**Skills level**: junior+, mid

**Mentors:** Lia Domide (lead), Romina Baila (backup)

**_All projects bellow can be tailored for different lenghts, as the features/pages can be build incrementally._**

**Comment by peeplika** [comment]
Helllo @liadomide 
Thank you for the update and for directing me to the correct thread for GSOC 2024. The project resonates with my interests, and I am excited to contribute. Could you provide guidance on the right approach for getting started or any resources? Additionally, I'd like to understand your expectations from contributors like me. I'm enthusiastic about the opportunity to contribute to GSoC.

Thanks
Priya

**Reply by liadomide** [reply]
Hi Priya,

From you, once the GSOC project gets published by INCF am,d Google, we expect:
- to comply with GSOC rules and submit a project proposal ; ideally you would share that proposal as a draft doc with us in advance, so we can give you early feedback;
- you prove in that proposal that you understand what jupyter lab is, what brings as base env for developing widgets in
- you prove that you understood and tried our current widgets https://github.com/the-virtual-brain/tvb-widgets/
- you try to identify what extra new 3D Head related widgets can be added: you alone can estimate how many such widgets you can build considering the project length that you choose

Let me know is anything gets unclear,
Lia.

**Comment by peeplika** [comment]
Thank you for clarifying. I'm in the process of familiarizing myself with the repository and widgets.I will work on understanding it thoroughly. I'll make sure to submit the proposal early, as suggested.

**Comment by peeplika** [comment]
Hello @liadomide,

I attempted to use the Widget 3D Head following the instructions provided here: https://wiki.ebrains.eu/bin/view/Collabs/tvb-widgets/Widget%203D/. However, it appears that the widget requires a Collab Auth Token, for which I need to create an EBRAINS account. Unfortunately, the account creation process asks for an institutional email, which I currently do not have access to.
Could you please advise on how to proceed in this situation? Is there an alternative method ?I have already wrote to their support team and am currently awaiting their response.

Thanks
Priya

**Reply by liadomide** [reply]
Hi Priya,
Indeed, some parts of tvb-widgets require an EBRAINS account, in particular all widgets with `*Browser` in their names.
But the functional visualization part should be fine without a token.
See here https://wiki.ebrains.eu/bin/view/Collabs/tvb-widgets/Widget%203D/ the cell with 
`widget = api.HeadWidget()` and bellow.
Let me know if you are still blocked after knowing these.

**Reply by peeplika** [reply]
Hii Lia,
I've been trying to run the widget following the ``api.HeadWidget()`` method, but it keeps giving this error:
``tvb/basic/readers.py:122: UserWarning: loadtxt: input contained no data: "<zipfile.ZipExtFile name='vertex_normals.txt' mode='r' compress_type=deflate>"
  array_result = numpy.loadtxt(file_stream, dtype=dtype, skiprows=skip_rows, usecols=use_cols)
Error displaying widget``
I've been trying to resolve it, but I was wondering if you have any clue about what might be going wrong.Any suggestion or insight would be really helpful to me.
Thanks!

**Reply by liadomide** [reply]
Hi Priya, 
In the python env where you run this, did you install tvb-data package?
You can do it either by `pip install tvb-data` or yb downloading and manually installing from [here](https://zenodo.org/records/10128131)

**Comment by peeplika** [comment]
Hi Lia,
Thanks for your response! I've confirmed the tvb-data package is installed, and some notebooks from [here](https://github.com/the-virtual-brain/tvb-widgets/tree/main/notebooks) are running well. However, the  Headwidget using `api.HeadWidget()` is still throwing an error in this line-`widget.add_datatype(face)`. I think it might be a path issue and I'm looking into it.

**Reply by liadomide** [reply]
I suggest we take this offline, as it gets too long for being useful for a larger crowd. 
We might come back here with just the solution.
Please email me the full stack trace and notebook in error at ldomide@gmail.com
